### PR TITLE
Customizable permalinks

### PR DESCRIPTION
**Pull request by martinklepsch**
Permalinks should be easily customizable. The idea here is that the `permalink` task gets a function that determines the permalink based on the metadata of a file. We could create a slug based on the title of a file but that is complex and actually we already have a perfect slug at hand: the filename itself.

In most cases though the filename contains dates and other unwanted things. Thats where the `slug` task comes in: it adds a `:slug` key to the metadata based on the result of a `slug-fn` on the filename. `slug-fn` can be overwritten via task options of course.

Let me know what you think

### CODE REVIEW

File: src/io/perun.clj [hunk 1]
```diff
@@ -74,18 +73,43 @@
...
+       (drop 3)
+       drop-last
+       (string/join "-")
+       string/lower-case))
+
+(deftask slug
+  "Adds :slug key to files metadata. Slug is derived from filename."
+  [s slug-fn SLUGFN sym "Function to build slug from filename"]
+  (boot/with-pre-wrap fileset
+    (let [slug-fn       (or slug-fn default-slug-fn)
```
Comment (Deraen) [inline src/io/perun.clj:87]: If slug-fn is a symbol it won't work as function?

Comment (martinklepsch) [inline src/io/perun.clj:87]: Oh right, that probably should be `code`.

**Comment by podviaznikov** [comment]
I'm going to go through this PR in few hours. Looks good so far!

File: src/io/perun.clj [hunk 2]
```diff
@@ -15,23 +16,23 @@
...
+
 (deftask markdown
-  "Parse markdown files"
-  [f create-filename CREATE_FILENAME str "Function that creates final target filename of the file"]
+  "Parse markdown files
+
+  This task will look for files ending with `md` or `markdown`
+  and add a `:content` key to their metadata containing the
+  HTML resulting from processing the markdown file's content"
+  []
```
Comment (podviaznikov) [inline src/io/perun.clj:34]: :+1: for removing `create-filename` from markdown task

File: src/io/perun.clj [hunk 3]
```diff
@@ -74,18 +73,43 @@
...
+       (string/join "-")
+       string/lower-case))
+
+(deftask slug
+  "Adds :slug key to files metadata. Slug is derived from filename."
+  [s slug-fn SLUGFN code "Function to build slug from filename"]
+  (boot/with-pre-wrap fileset
+    (let [slug-fn       (or slug-fn default-slug-fn)
+          files         (:metadata (meta fileset))
+          updated-files (into {}
```
Comment (podviaznikov) [inline src/io/perun.clj]: You can use `https://github.com/hashobject/perun/blob/master/src/io/perun/core.clj#L36` function here. Will be a little bit shorter

File: src/io/perun.clj [hunk 4]
```diff
@@ -138,11 +162,11 @@
         (commit fileset tmp)))))
 
 (def ^:private +render-defaults+
-  {:target "public"})
+  {:out-dir "public"})
```
Comment (podviaznikov) [inline src/io/perun.clj:165]: :+1: for renaming

File: src/io/perun.clj [hunk 5]
```diff
@@ -151,22 +175,22 @@
         (doseq [file files]
           (let [render-fn (:renderer options)
                 html (render-fn file)
-                page-filepath (str (:target options) "/"
-                                   (or (:filepath file)
+                page-filepath (str (:out-dir options) "/"
+                                   (or (:permalink file)
```
Comment (podviaznikov) [inline src/io/perun.clj:179]: :+1: for making permalink prop

**Comment by podviaznikov** [comment]
Looks much cleaner now. I like how each task does one job now.

**Pull request merged**